aesd-char-driver/main.c

```diff
@@ -71,6 +71,179 @@ int aesd_release(struct inode *inode, struct file *filp)
     return 0;
 }
 
+/**
+ * "The llseek method is used to change the current read/write position
+ * in a file, and the new position is returne as a (positive) return value.
+ * The loff_t parameter is a "long offset" and is at least 64 bits wide 
+ * even on 32-bit platforms. Errors are signaled by a negative return value.
+ * If this function pointer is NULL, seek calls will modify the position
+ * counter in the file structure in potentially unpredictable ways."
+ * Linux Device Drivers Chapter 3, J. Corbet, A Rubini, G. Kroah-Hartman
+ */
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+    ssize_t retval;
+    loff_t fsl;
+    loff_t circbuffsize=0;
+    struct aesd_dev *dev = NULL;
+    struct aesd_buffer_entry *entry = NULL;
+    uint8_t stread;
+    if (filp == NULL)
+    {
+        retval = -EINVAL;
+        PDEBUG("file pointer filp NULL, returning %zu", retval);
+        goto errout;
+    }
+
+    dev = filp->private_data;
+
+    PDEBUG("entered seek with offset = %lld and whence = %d", offset, whence);
+
+    if (mutex_lock_interruptible(&dev->bufferlock))
+    {
+        retval = -ERESTARTSYS;
+        PDEBUG("aesd_llseek: mutex lock interruptible , returning %zu", retval);
+        goto errout;
+    }
+
+    AESD_CIRCULAR_BUFFER_FOREACH(entry, &(aesd_device.circular), stread)
+    {
+        if(entry->buffptr != NULL)
+            circbuffsize += entry->size;
+    }
+
```

Comment:
@sota6640 could release the mutex here since you are no longer accessing the device at this point.